### PR TITLE
MONGOCRYPT-474 Support converting Int32 and Int64 for Range Indexes

### PR DESCRIPTION
**Pull request by kevinAlbs**
# Summary
- Support converting Int32 and Int64 for Range Indexes.
- Add macro to enable `-Wconversion`.
- Add assertion comparisons for other integral types, (e.g. `ASSERT_CMPUINT32`).

# Background & Motivation

Range Index search requires encoding supported numeric types as unsigned integers. The encoding process is described in https://docs.google.com/document/d/1ABDt7WqVasSGfS0tYBTVWlbxZr9IBN_aZneitPRg2vI/edit#heading=h.a6eu0rvjme13.

This is a partial translation of the work done in https://github.com/mongodb/mongo/commit/a951cece668963881c5ff6dce0b0199782f9cf53. It excludes double. The `min` and `max` bounds for double are not implemented on the server yet.

### CODE REVIEW

File: src/mc-range-encoding.c [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+/**
+ * Encode a signed 32-bit integer as an unsigned 32-bit integer by adding 2^31.
+ * Some documentation references this as making the value "unbiased".
+ */
+static uint32_t
+encodeInt32 (int32_t v)
+{
+   // Shift the int32_t range [-2^31, 2^31 - 1] to the uint32_t range [0, 2^32].
+   // new_zero is the mapped 0 value.
+   uint32_t new_zero = (1U << 31);
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/mc-range-encoding.c]: Done.

File: src/mc-range-encoding.c [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+/**
+ * Encode a signed 64-bit integer as an unsigned 64-bit integer by adding 2^63.
+ * Some documentation references this as making the value "unbiased".
+ */
+static uint64_t
+encodeInt64 (int64_t v)
+{
+   // Shift the int64_t range [-2^63, 2^63 - 1] to the uint64_t range [0, 2^64].
+   // new_zero is the mapped 0 value.
+   uint64_t new_zero = (1ULL << 63);
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/mc-range-encoding.c]: Done.

File: test/test-mc-range-encoding.c [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+typedef struct {
+   mc_getTypeInfo32_args_t args;
+   mc_OSTType_Int32 expect;
+   const char *expectError;
+} Int32Test;
+
+static void
+_test_RangeTest_Encode_Int32 (_mongocrypt_tester_t *tester)
+{
+   mongocrypt_status_t *status;
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply test/test-mc-range-encoding.c]: Done.

File: test/test-mc-range-encoding.c [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+   Int32Test tests[] = {
+      /* Test cases copied from server Int32_NoBounds test ... begin */
+      {.args = {.value = 2147483647},
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply test/test-mc-range-encoding.c]: Done. 

`UINT32_C` seems more readable than the `U` suffix. Out of curiosity, do you know if there are other reasons to prefer the macro over the suffix?

File: test/test-mc-range-encoding.c [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+typedef struct {
+   mc_getTypeInfo64_args_t args;
+   mc_OSTType_Int64 expect;
+   const char *expectError;
+} Int64Test;
+
+static void
+_test_RangeTest_Encode_Int64 (_mongocrypt_tester_t *tester)
+{
+   mongocrypt_status_t *status;
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply test/test-mc-range-encoding.c]: Done.

File: test/test-mongocrypt-assert.h [hunk 6]
```diff
@@ -136,16 +136,35 @@
...
-      int _a_int = _a;                                                  \
-      int _b_int = _b;                                                  \
-      if (!(_a_int _operator _b_int)) {                                 \
-         TEST_ERROR (                                                   \
-            "comparison failed: %d %s %d", _a_int, #_operator, _b_int); \
-      }                                                                 \
+#define ASSERT_CMP_HELPER(_a, _operator, _b, fmt, type)      \
+   do {                                                      \
+      type _a_value = _a;                                    \
+      type _b_value = _b;                                    \
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply test/test-mongocrypt-assert.h]: Done.

File: test/test-mongocrypt-assert.h [hunk 7]
```diff
@@ -136,16 +136,35 @@
...
    } while (0);
 
+#define ASSERT_CMPINT(a, eq, b) ASSERT_CMP_HELPER (a, eq, b, "d", int)
+#define ASSERT_CMPUINT(a, eq, b) ASSERT_CMP_HELPER (a, eq, b, "u", unsigned int)
+#define ASSERT_CMPLONG(a, eq, b) ASSERT_CMP_HELPER (a, eq, b, "ld", long)
+#define ASSERT_CMPULONG(a, eq, b) \
+   ASSERT_CMP_HELPER (a, eq, b, "lu", unsigned long)
+#define ASSERT_CMPINT32(a, eq, b) ASSERT_CMP_HELPER (a, eq, b, PRId32, int32_t)
+#define ASSERT_CMPINT64(a, eq, b) ASSERT_CMP_HELPER (a, eq, b, PRId64, int64_t)
+#define ASSERT_CMPUINT16(a, eq, b) ASSERT_CMP_HELPER (a, eq, b, "hu", uint16_t)
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply test/test-mongocrypt-assert.h]: Done.

File: test/test-mongocrypt-assert.h [hunk 8]
```diff
@@ -136,16 +136,35 @@
...
+#define ASSERT_CMPUINT32(a, eq, b) \
+   ASSERT_CMP_HELPER (a, eq, b, PRIu32, uint32_t)
+#define ASSERT_CMPUINT64(a, eq, b) \
+   ASSERT_CMP_HELPER (a, eq, b, PRIu64, uint64_t)
+#define ASSERT_CMPSIZE_T(a, eq, b) ASSERT_CMP_HELPER (a, eq, b, "zu", size_t)
+#define ASSERT_CMPSSIZE_T(a, eq, b) ASSERT_CMP_HELPER (a, eq, b, "zd", ssize_t)
+#define ASSERT_CMPDOUBLE(a, eq, b) ASSERT_CMP_HELPER (a, eq, b, "f", double)
+#define ASSERT_CMPVOID(a, eq, b) ASSERT_CMP_HELPER (a, eq, b, "p", void *)
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply test/test-mongocrypt-assert.h]: Done.

File: test/test-mongocrypt-assert.h [hunk 9]
```diff
@@ -136,16 +136,36 @@
...
 
+#define ASSERT_CMPINT(a, eq, b) ASSERT_CMP_HELPER (a, eq, b, "d", int)
+#define ASSERT_CMPUINT(a, eq, b) ASSERT_CMP_HELPER (a, eq, b, "u", unsigned int)
+#define ASSERT_CMPLONG(a, eq, b) ASSERT_CMP_HELPER (a, eq, b, "ld", long)
+#define ASSERT_CMPULONG(a, eq, b) \
+   ASSERT_CMP_HELPER (a, eq, b, "lu", unsigned long)
+#define ASSERT_CMPINT32(a, eq, b) ASSERT_CMP_HELPER (a, eq, b, PRId32, int32_t)
+#define ASSERT_CMPINT64(a, eq, b) ASSERT_CMP_HELPER (a, eq, b, PRId64, int64_t)
+#define ASSERT_CMPUINT16(a, eq, b) \
+   ASSERT_CMP_HELPER (a, eq, b, PRId16, uint16_t)
```
Comment (eramongodb) [inline test/test-mongocrypt-assert.h:159]:
```suggestion
#define ASSERT_CMPUINT16(a, eq, b) \
   ASSERT_CMP_HELPER (a, eq, b, PRIu16, uint16_t)
```

File: test/test-mc-range-encoding.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (eramongodb) [reply test/test-mc-range-encoding.c]: `UINT*_C` ensures that regardless of size of either `uint*_t` or `int`/`long`/`long long` that the literal will be in a manner that is representable by `uint*_t`. There is a motivating example presented in the [proposal for fixed-width integer user-defined literals for C++](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1280r2.html):
```cpp
#include <cstdint>

void foo(uint32_t);
void foo(uint64_t);

void test()
{
    // change this to 10ul and it fails to compile on MSVC
    // change this to 10ull and it fails to compile on GCC and Clang
    // The only safe thing to do is to use the UINTXX_C macros.
    auto x = 10; 
    foo(x);
}
```
Although C does not have function overloads, the principle is still the same: ensuring type-and-literal pairings are compatible and portable across implementations. Admittedly it is not as strictly necessary for C unless the literal is being passed to e.g. `printf`:
```c
printf("%" PRIu32 "\n", 123ul); // Cannot portably assume underlying type.
printf("%" PRIu32 "\n", UINT32_C(123)); // OK
```
However I still find it valuable to see the type in the literal expression itself to verify the code is aware of and consistent in its use of fixed-width integral types and values as a "best practices" (some may disagree on this point and consider this redundant/excessive).

[the remark this answers is not in the document]
Comment (kevinAlbs) [reply test/test-mc-range-encoding.c]: The proposal is interesting. Thank you for sharing.

**Review by eramongodb — APPROVED** [review]

**Review by vector-of-bool — APPROVED** [review]
Sorry for the delay on this. Following all of Ezra's comments, LGTM!

**Pull request merged**